### PR TITLE
:technologist: `Edge` and `Graph` --- Interfaces 

### PR DESCRIPTION
**Pull request by jameshughes89**
### What
1. Define an edge interface
2. Define a graph interface

### How
I used [Google's Guava's Graph stuff as a reference](https://guava.dev/releases/snapshot/api/docs/com/google/common/graph/Graph.html).  

### Additional Notes
The idea is that these could be extended to be weighted or directed. 

Notice that there are no add or remove methods. The plan is to have `Graph` objects be immutable and be created with a static factory that takes a list of `Edge` objects or something. 

The `getNodes()` and `getEdges()` do not throw exceptions as I figure a graph can be empty and should therefore return empty sets.

### CODE REVIEW

**Review by twentylemon — APPROVED** [review]
> Notice that there are no add or remove methods.

Immutable data structures still tend to have mutation operations. That's kind of their whole jam.

**Comment by jameshughes89** [comment]
> > Notice that there are no add or remove methods.
> 
> Immutable data structures still tend to have mutation operations. That's kind of their whole jam.

You're saying, you think I'm gonna' want add & remove in regardless?

**Comment by twentylemon** [comment]
I think I'm saying to basically complicate by making it mutable. For better or worse, mutable data structures are the norm and the point here is to show other data structures. Immutable data structures include a fair bit of logic to allow for mutations, which would muddy the actual lesson. Though, persistent data structures may be a good assignment.

**Comment by jameshughes89** [comment]
I based the interface on [Google's stuff](https://guava.dev/releases/23.0/api/docs/com/google/common/graph/MutableGraph.html)

1. I included `addNode`, `removeNode`, `addEdge`, and `removeEdge` 
2. For `addNode`, I opted to have it throw an exception if the node already exists. Feels like silently ignoring would work, but idk, exception feels better and more consistent with how things have been done so far. 
3. For `addEdge`, I opted to throw an exception if the nodes do not exist. Could have also just added the nodes if they do not exist, but similar to above, exception seems better since, how the F do I know wtf you want to do.

**Comment by twentylemon** [comment]
> 2\. For `addNode`, I opted to have it throw an exception if the node already exists.

Boolean return seems more sensible. Graphs are just sets, and java's `Set#add` returns boolean.

Throwing for an invalid edge though, seems reasonable.

**Comment by jameshughes89** [comment]
> > 2. For `addNode`, I opted to have it throw an exception if the node already exists.
> 
> Boolean return seems more sensible. Graphs are just sets, and java's `Set#add` returns boolean.
> 
> Throwing for an invalid edge though, seems reasonable.

Changed 
- `addNode` to return `false` if node already exists and updated javadoc accordingly. 
- `addEdge` to return `false` if the edge exists and updated javadoc accordingly

File: src/main/java/Graph.java [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+     * @return True if the node was added successfully, false otherwise
+     */
+    boolean addNode(T node);
+
+    /**
+     * Removes the specified node from the graph. All edges incident on the node are also removed.
+     *
+     * @param node The node to be removed from the graph
+     * @return True if the node was removed successfully
+     * @throws NoSuchElementException If the specified node does not exist within the graph
```
Comment (twentylemon) [inline src/main/java/Graph.java]: exception here seems nonsensical if the return value already encodes that information

Comment (jameshughes89) [reply src/main/java/Graph.java]: The idea was to follow what was done with the other data structures in the previous course, but upon looking at the Java docs, I see that what was done with the previous data structures was perhaps atypical. For example, the `Bag` data structures throws an exception if removing a nonexistent element. Checking [Java's list ](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/util/List.html), which was the basis of the bag data structure, does not throw an exception. 

My guess is that I followed the removes from the stack and queue (pop, dequeue) that threw an exception when empty, but obviously that's a different case. 

Long story short, good call; I'll make the change here and will man an issue for 102's content to update the removes to NOT throw an exception when the element does not exist.

**Comment by jameshughes89** [comment]
> Throwing for an invalid edge though, seems reasonable.

Although this is likely your point, to just have it stated, the motivation for having the exception for the add edge is

**Review by twentylemon — APPROVED** [review]
